File: src/Python/extract_matrix.py
###########################################################
# @Author: HoodUSSEnterprise
# @Date: 2026-06-22 13:20:38
# @LastEditors: HoodUSSEnterprise
# @LastEditTime: 2026-06-22 13:21:04
# @FilePath: \asm_matrix_benchmark\src\Python\extract_matrix.py
# @Description:extract submatrix python code
###########################################################

from typing import List, Union

from .base_matrix import Matrix


###########################################################
# @description: extract a row from matrix as new Matrix (1×n)
# @param {Matrix} m
# @param {int} row_idx
# @return {Matrix}
###########################################################
def extract_row(m: "Matrix", row_idx: int) -> "Matrix | None":
    if m is None:
        print("Invalid param")
        return None
    if row_idx < 0 or row_idx >= m.rows:
        print(f"Row index {row_idx} out of range [0, {m.rows - 1}]")
        return None

    start = row_idx * m.cols
    data = m.data[start : start + m.cols]
    return Matrix(1, m.cols, data)


###########################################################
# @description: extract a column from matrix as new Matrix (m×1)
# @param {Matrix} m
# @param {int} col_idx
# @return {Matrix}
###########################################################
def extract_col(m: "Matrix", col_idx: int) -> "Matrix | None":
    if m is None:
        print("Invalid param")
        return None
    if col_idx < 0 or col_idx >= m.cols:
        print(f"Column index {col_idx} out of range [0, {m.cols - 1}]")
        return None

    data: List[Union[int, float]] = [
        m.data[i * m.cols + col_idx] for i in range(m.rows)
    ]
    return Matrix(m.rows, 1, data)


###########################################################
# @description: extract a submatrix block
# @param {Matrix} m
# @param {int} r_start : row start (inclusive)
# @param {int} r_end   : row end (exclusive)
# @param {int} c_start : col start (inclusive)
# @param {int} c_end   : col end (exclusive)
# @return {Matrix}
###########################################################
def extract_submatrix(
    m: "Matrix", r_start: int, r_end: int, c_start: int, c_end: int
) -> "Matrix | None":
    if m is None:
        print("Invalid param")
        return None

    if r_start < 0 or r_end > m.rows or r_start >= r_end:
        print(f"Invalid row range [{r_start}, {r_end})")
        return None
    if c_start < 0 or c_end > m.cols or c_start >= c_end:
        print(f"Invalid col range [{c_start}, {c_end})")
        return None

    rows_out = r_end - r_start
    cols_out = c_end - c_start
    data: List[Union[int, float]] = [0 for _ in range(rows_out * cols_out)]

    for i in range(rows_out):
        for j in range(cols_out):
            data[i * cols_out + j] = m.data[(r_start + i) * m.cols + (c_start + j)]

    return Matrix(rows_out, cols_out, data)

File: src/C++/float/extract_matrix.cpp
/************************************************************
@Author: HoodUSSEnterprise
@Date: 2026-06-23 17:39:10
@LastEditors: HoodUSSEnterprise
@LastEditTime: 2026-06-26 09:04:19
@FilePath: \asm_matrix_benchmark\src\C++\float\extract_matrix.cpp
@Description: extract matrix c++ code
*************************************************************/

#include "matrix_float_cpp.h"
#include <stdexcept>

/***********************************************************
@description:  Extracts a single row from the matrix as a new 1-row matrix
@param {size_t} row_idx
@return {*}
*************************************************************/
MatrixFloat MatrixFloat::extract_row(size_t row_idx) const
{
    if (row_idx >= rows)
    {
        throw std::invalid_argument("extract_row: index out of range");
    }

    // Copy the entire row into a 1 x cols matrix
    MatrixFloat res(1, cols);
    for (size_t j = 0; j < cols; j++)
    {
        res.data[j] = data[row_idx * cols + j];
    }
    return res;
}

/***********************************************************
@description: Extracts a single col from the matrix as a new 1-row matrix
@param {size_t} col_idx
@return {*}
*************************************************************/
MatrixFloat MatrixFloat::extract_col(size_t col_idx) const
{
    if (col_idx >= cols)
    {
        throw std::invalid_argument("extract_col: index out of range");
    }

    // Copy the entire column into a rows x 1 matrix
    MatrixFloat res(rows, 1);
    for (size_t i = 0; i < rows; i++)
    {
        res.data[i] = data[i * cols + col_idx];
    }
    return res;
}

/***********************************************************
@description: Extracts a submatrix (contiguous block) from the matrix using half-open interval [r_start, r_end) x [c_start, c_end)
@param {size_t} r_start
@param {size_t} r_end
@param {size_t} c_start
@param {size_t} c_end
@return {*}
@throws {std::invalid_argument} If any of the following conditions are met:
        - r_start >= r_end (empty row range)
        - c_start >= c_end (empty column range)
        - r_end > rows (row range exceeds matrix bounds)
        - c_end > cols (column range exceeds matrix bounds)
*************************************************************/
MatrixFloat MatrixFloat::extract_submatrix(size_t r_start, size_t r_end, size_t c_start, size_t c_end) const
{
    if (r_start >= r_end || c_start >= c_end || r_end > rows || c_end > cols)
    {
        throw std::invalid_argument("extract_submatrix: invalid range");
    }

    size_t r_out = r_end - r_start;
    size_t c_out = c_end - c_start;
    MatrixFloat res(r_out, c_out);
    // Copy the contiguous block element by element
    for (size_t i = 0; i < r_out; i++)
    {
        for (size_t j = 0; j < c_out; j++)
        {
            res.data[i * c_out + j] = data[(r_start + i) * cols + (c_start + j)];
        }
    }
    return res;
}
